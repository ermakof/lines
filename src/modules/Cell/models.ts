export type TPoss = [number, number];

export enum ICellInfo {
  empty = 0,
  red = 1,
  yellow = 2,
  green = 3,
}

export interface CellArray {
  size: number;
  data: ICellInfo[];
}
