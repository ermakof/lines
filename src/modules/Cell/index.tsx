import React from 'react';
import { useDispatch } from 'react-redux';
import styled from '@emotion/styled';
import { CELL_SIZE } from '@src/modules/Cell/constatnts';
import { actions } from '@src/App/appSlice';

interface IContainer {
  isRight?: boolean;
  isLeft?: boolean;
  isBottom?: boolean;
}
const Container = styled.div<IContainer>`
  display: block;
  width: ${({ isRight }) => CELL_SIZE - (isRight ? 2 : 1)}px;
  height: ${({ isBottom }) => CELL_SIZE - (isBottom ? 2 : 1)}px;
  text-align: center;
  border-top: 1px solid #ddd;
  border-left: 1px solid #ddd;
  border-right: ${({ isRight }) => `1px solid ${isRight ? '#ddd' : ''};`}
  border-bottom: ${({ isBottom }) => `1px solid ${isBottom ? '#ddd' : ''};`}
  float: left;
  ${({ isLeft }) => (isLeft ? 'clear: both;' : '')}
  transition-property: background;
  transition-duration: 0.3s;
  transition-timing-function: ease-out;
  background: #555;
  background: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;  
`;

interface IContent {
  selected?: boolean;
  highlighted?: string;
}
const Content = styled.div<IContent>`
  cursor: pointer;
  background: #ffff00;
  border-radius: 50%;
  background: ${({ selected, highlighted }) =>
    !highlighted ? '#ffff0070' : selected ? '#ffff00' : highlighted};
  border-radius: 50%;
  width: 30px;
  height: 30px;
  margin: 5px;
  border: 0.6vh solid transparent;
  border-color: ${({ selected }) => (selected ? '#555' : '#9e9e9e')};

  ${({ highlighted }) => (!highlighted ? ':hover {background: #ffff00;}' : '')}
`;

export interface CellProps {
  num: number;
  isFilled?: number;
  isSelected: boolean;
  isLeft?: boolean;
  isRight?: boolean;
  isBottom?: boolean;
  highlighted: string;
}

const Cell: React.FC<CellProps> = ({
  num,
  isFilled = 0,
  isSelected = false,
  isLeft = true,
  isRight = true,
  highlighted = '',
}) => {
  const dispatch = useDispatch();

  const handleSelectCell = (event: React.MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    if (!highlighted) {
      dispatch(actions.setSelectedCell(!isSelected ? num : undefined));
    }
  };

  const handleSelectContainer = (event: React.MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    if (!isSelected && !highlighted) {
      dispatch(actions.startGameSteps(num));
    }
  };

  return (
    <Container
      role={`cellContainer-${num}`}
      isRight={isRight}
      isLeft={isLeft}
      onClick={handleSelectContainer}
    >
      {!!isFilled && (
        <Content
          role={`cellContent-${num}`}
          selected={isSelected}
          onClick={handleSelectCell}
          highlighted={highlighted}
        />
      )}
    </Container>
  );
};

export default Cell;
