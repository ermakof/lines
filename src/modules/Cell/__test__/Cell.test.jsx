import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import { useDispatch } from 'react-redux';

import Cell from '@src/modules/Cell';

const mockStore = configureMockStore();
let store;

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('Cell', () => {
  beforeEach(() => {
    store = mockStore({
      gameFieldData: [1, 1, 1, 1, 1, 1, 1, 1, 1],
      gameLevel: '1',
      gameFieldPercentFilled: 100,
    });
  });

  it('renders <Cell> using renderer cell with value=2', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Cell num={2} isFilled={1} isSelected={true} filled={1} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const cellContainer = screen.getByRole(/cellContainer/gi);
    expect(cellContainer).toBeInTheDocument();
    const cellContent = screen.getByRole(/cellContent-2/gi);
    expect(cellContent).toBeInTheDocument();
  });

  it('Click on content cell', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    render(
      <Provider store={store}>
        <Cell num={2} isFilled={1} isSelected={true} highlighted={''} filled={1} />
      </Provider>
    );
    const cellContent = screen.getByRole(/cellContent-2/i);
    fireEvent.click(cellContent);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(cellContent).toHaveStyle('background: #ff0000');
  });

  it('Click on container cell', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    render(
      <Provider store={store}>
        <Cell num={2} isFilled={1} isSelected={false} isOutdated={false} filled={1} />
      </Provider>
    );
    const cellContainer = screen.getByRole(/cellContainer-2/i);
    fireEvent.click(cellContainer);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });
});
