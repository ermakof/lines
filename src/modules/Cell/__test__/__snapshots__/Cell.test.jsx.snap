// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders <Cell> using renderer cell with value=2 1`] = `
<DocumentFragment>
  <div
    class="css-upvruk"
    role="cellContainer-2"
  >
    <div
      class="css-vba8le"
      role="cellContent-2"
    />
  </div>
</DocumentFragment>
`;
